Add thai keyboard distance

ppirch opened the pull request:
### What does this changes

Add function for calculate euclidean distance between two Thai characters on QWERTY Layout Keyboard

ppirch committed:
Add test for thai keyboard distance

ppirch committed:
Implement thai keyboard distance

pep8speaks commented:
Hello @ppirch! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:







There are currently no PEP 8 issues detected in this Pull Request. Cheers! :beers: 

##### Comment last updated at 2021-01-04 02:18:48 UTC

ppirch committed:
remove trailing whitespace

ppirch committed:
split line

ppirch committed:
remove whitespace

ppirch committed:
remove trailing whitespace

bact commented on pythainlp/util/keyboard.py:
```diff
@@ -103,6 +103,20 @@
+  "ฤ", "ฆ", "ฏ", "โ", "ฌ", "็", "๋", "ษ", "ศ", "ซ", ".",
+  "(", ")", "ฉ", "ฮ", "ฺ", "์", "?", "ฒ", "ฬ", "ฦ"
+]
+
```
Do we suppose to have 4 rows for each keyboard layout here?

I'm not entirely sure about the current logic, but it seems to me these TH_QWERTY and TH_SHIFT_QWERTY have only one row (they're all 1-dimension array).

ppirch replied:
I'm sorry it was my fault. It should be 2-dimension array. I forgot to put [] between each row.

bact commented:
Thanks for great contribution.
Do you mind to change the name of the layout from QWERTY to "Kedmanee" (or TIS 820 / TIS 820-2531 / TIS 820-2538, if that matter) ? 
https://en.wikipedia.org/wiki/Thai_Kedmanee_keyboard_layout

ppirch committed:
Change the name of the layout from QWERTY to TIS_820_2538

ppirch commented:
> Thanks for great contribution.
> Do you mind to change the name of the layout from QWERTY to "Kedmanee" (or TIS 820 / TIS 820-2531 / TIS 820-2538, if that matter) ?
> https://en.wikipedia.org/wiki/Thai_Kedmanee_keyboard_layout

I agree to that suggestion. 
I changed the name from QWERTY to TIS 820-2538.

wannaphong commented on pythainlp/util/keyboard.py:
```diff
@@ -103,6 +103,20 @@
 TH_EN_TRANSLATE_TABLE = str.maketrans(TH_EN_KEYB_PAIRS)
 
+TIS_820_2538 = [
+  "-", "ๅ", "/", "_", "ภ", "ถ", "ุ", "ึ", "ค", "ต", "จ", "ข", "ช",
```
`"-", "ๅ", "/", "_"` -> `"_", "ๅ", "/", "-"`

ppirch replied:
It's fixed.

ppirch committed:
Fix layout

wannaphong committed:
Merge branch 'dev' into add-thai-keyboard-distance

wannaphong commented:
I have checked the algorithm. It's passed. `get_char_coord` output is x,y. The formula is ![formula](https://render.githubusercontent.com/render/math?math={\displaystyle%20d(x,y)={\sqrt%20{(x_{1}-x_{2})^{2}\plus(y_{1}-y_{2})^{2}}}.}&mode=inline)

bact committed:
Fix keyboard layout array

- Fix: According to the current distance algorithm, the keyboard layout should be a 2D array (in this case, 4 rows x N columns)
  - Try "๘"-"๙" and "๙"-"๐", they should have different distance values.
- Fix: The proposed layout is actually a modified TIS 820-2531 (not 2538 one, see docstring)

bact committed:
Add keyboard distance test

Thai digit zero (๐) shouldn't be "next to" Thai digit nine (๙).
Their distance should be different from the distance between, say, Thai digit eight (๘) and Thai digit nine (๙).

bact reviewed (approved):
@ppirch I have committed my proposed code, please kindly check if I understand your algorithm correctly.

I basically changed the keyboard layout array from 1D to 2D.
Since the distance between Thai digit eight and nine should be different from the distance between Thai digit nine and zero.
These two pairs are both next to each other in 1D array, but it's not the case for 2D keyboard layout.

A new test case also added to illustrated this.

I also changed the constant names to refer to a modified TIS 820 2531 instead. Since that's more reflected the content inside the array.

bact committed:
Fix PEP8 whitespace after comma

bact committed:
Update thai_keyboard_dist comment

bact committed:
Update keyboard.py

bact commented:
Sorry, I made a mistake in reverting the merge. Reverting the revert, and this PR is safely restored.

Thanks @ppirch again for the code.
